File: frontend/src/services/performance.ts
// Serviço do módulo "Minha Equipe" (Performance de Equipes).
// Self-contained (tipos + chamadas) pra não inflar o api.ts.

import { apiFetch } from "@/lib/api-client";

const BASE = "/api/v1/performance";

async function json<T>(res: Response): Promise<T> {
  if (!res.ok) {
    let detail = `Erro ${res.status}`;
    try {
      detail = (await res.json())?.detail || detail;
    } catch {
      /* ignore */
    }
    throw new Error(detail);
  }
  return res.json() as Promise<T>;
}

export type Categoria = "operacional" | "profundo" | "ruido";

export interface PessoaMetrica {
  id: number;
  nome: string;
  cargo: string | null;
  squad: string | null;
  posicao: string | null;
  concluido: number;
  dias_ativos: number;
  throughput_dia: number;
  no_prazo_pct: number | null;
  cycle_dias: number | null;
  backlog: number;
  operacional_n: number;
  profundo_n: number;
  ruido_n: number;
}

export interface EquipeKpis {
  concluido: number;
  backlog: number;
  pessoas_ativas: number;
  pessoas_total: number;
  no_prazo_pct: number | null;
}

export interface EquipeResponse {
  periodo_dias: number;
  kpis: EquipeKpis;
  pessoas: PessoaMetrica[];
}

export interface MixItem {
  subtipo: string;
  categoria: Categoria;
  volume: number;
  cycle_dias: number | null;
  no_prazo_pct: number | null;
  tempo_tarefa_seg: number | null;
}

export interface RitmoOcio {
  volume: number;
  cadencia_seg: number | null;
  ocio_pct: number | null;
  dias: number;
  oper_share: number | null;
  inicio_h: number | null;
  fim_h: number | null;
}

export interface PassadoKpis {
  concluido: number;
  dias_ativos: number;
  throughput_dia: number;
  no_prazo_pct: number | null;
  cycle_dias: number | null;
}

export interface PendenteTipo {
  subtipo: string;
  categoria: Categoria;
  total: number;
  atrasado: number;
}

export interface UrgenteItem {
  subtipo: string;
  prazo: string | null;
  dias: number | null;
  atrasado: boolean;
  cnj: string | null;
  pasta: string | null;
}

export interface PessoaDetalhe {
  pessoa: { id: number; nome: string; cargo: string | null; squad: string | null; posicao: string | null };
  periodo_dias: number;
  passado: { kpis: PassadoKpis; ritmo: RitmoOcio; mix: MixItem[] };
  futuro: {
    pendente: number;
    atrasado: number;
    sem_prazo: number;
    por_tipo: PendenteTipo[];
    urgentes: UrgenteItem[];
  };
}

export interface TipoItem {
  subtipo: string;
  categoria: Categoria;
  volume: number;
  pessoas: number;
  cycle_dias: number | null;
  densidade: number | null;
}

export async function getEquipe(team: string, days = 30, cargo?: string): Promise<EquipeResponse> {
  const qs = new URLSearchParams({ team, days: String(days) });
  if (cargo) qs.set("cargo", cargo);
  return json(await apiFetch(`${BASE}/equipe?${qs.toString()}`));
}

export async function getCargos(team: string): Promise<string[]> {
  const r = await json<{ cargos: string[] }>(await apiFetch(`${BASE}/cargos?team=${team}`));
  return r.cargos;
}

export async function getPessoa(id: number, team: string, days = 30): Promise<PessoaDetalhe> {
  return json(await apiFetch(`${BASE}/pessoa/${id}?team=${team}&days=${days}`));
}

export async function getTipos(team: string, days = 30): Promise<TipoItem[]> {
  const r = await json<{ tipos: TipoItem[] }>(await apiFetch(`${BASE}/tipos?team=${team}&days=${days}`));
  return r.tipos;
}

export interface VazaoItem {
  id: number;
  nome: string;
  cargo: string | null;
  concluido: number;
  throughput_dia: number;
}

export interface BacklogItem {
  id: number;
  nome: string;
  cargo: string | null;
  backlog: number;
  atrasado: number;
}

export interface JornadaItem {
  id: number;
  nome: string;
  cargo: string | null;
  inicio_h: number;
  fim_h: number;
  hands_on_h: number;
  ocio_pct: number | null;
  dias: number;
  oper_share: number;
}

export interface TopTipoItem {
  subtipo: string;
  categoria: Categoria;
  volume: number;
  pendente: number;
  atrasado: number;
}

export interface DashboardData {
  periodo_dias: number;
  kpis: { atrasado_total: number; backlog_total: number };
  vazao: VazaoItem[];
  backlog: BacklogItem[];
  jornada: JornadaItem[];
  top_tipos: TopTipoItem[];
}

export async function getDashboard(team: string, days = 30): Promise<DashboardData> {
  return json(await apiFetch(`${BASE}/dashboard?team=${team}&days=${days}`));
}

export async function downloadExport(params: {
  escopo: "atrasado" | "pendente" | "concluido";
  team: string;
  days: number;
  pessoa_id?: number;
  subtipo?: string;
}): Promise<void> {
  const qs = new URLSearchParams({ escopo: params.escopo, team: params.team, days: String(params.days) });
  if (params.pessoa_id) qs.set("pessoa_id", String(params.pessoa_id));
  if (params.subtipo) qs.set("subtipo", params.subtipo);
  const res = await apiFetch(`${BASE}/export?${qs.toString()}`);
  if (!res.ok) throw new Error(`Erro ${res.status} ao exportar`);
  const blob = await res.blob();
  const url = URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.href = url;
  a.download = `minha-equipe-${params.escopo}.xlsx`;
  document.body.appendChild(a);
  a.click();
  a.remove();
  URL.revokeObjectURL(url);
}

// Baixa o PDF (download via <a>) em vez de window.open: o open roda DEPOIS dos
// ~20s de geração (Sonnet), fora do gesto do usuário, e o navegador bloqueia o
// popup silenciosamente. O download por <a download> funciona após o await.
async function fetchPdfAndDownload(path: string, filename: string): Promise<void> {
  const res = await apiFetch(path);
  if (!res.ok) throw new Error(`Erro ${res.status} ao gerar o relatório`);
  const blob = await res.blob();
  const url = URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.href = url;
  a.download = filename;
  document.body.appendChild(a);
  a.click();
  a.remove();
  setTimeout(() => URL.revokeObjectURL(url), 10000);
}

export async function abrirRelatorioSetor(days: number): Promise<void> {
  return fetchPdfAndDownload(`${BASE}/relatorio-setor?days=${days}`, "relatorio-minha-equipe-setor.pdf");
}

export async function abrirRelatorioPessoa(id: number, days: number): Promise<void> {
  return fetchPdfAndDownload(`${BASE}/pessoa/${id}/relatorio?days=${days}`, `raio-x-pessoa-${id}.pdf`);
}

// ── Relatórios como job persistente ──
export interface RelatorioItem {
  id: number;
  tipo: string;
  label: string;
  days: number;
  status: "processando" | "pronto" | "erro";
  erro: string | null;
  criado_em: string | null;
  concluido_em: string | null;
}

export async function criarRelatorio(
  tipo: "setor" | "pessoa",
  team: string,
  days: number,
  pessoa_id?: number,
): Promise<{ id: number; label: string; status: string }> {
  return json(
    await apiFetch(`${BASE}/relatorios`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ tipo, team, days, pessoa_id }),
    }),
  );
}

export async function listarRelatorios(): Promise<RelatorioItem[]> {
  const r = await json<{ items: RelatorioItem[] }>(await apiFetch(`${BASE}/relatorios`));
  return r.items;
}

export async function downloadRelatorioById(id: number): Promise<void> {
  return fetchPdfAndDownload(`${BASE}/relatorios/${id}/download`, `relatorio-minha-equipe-${id}.pdf`);
}

// ── Ingestão dos dados (download do relatório do L1) ──
export interface SyncStatus {
  last_sync: {
    ok: boolean;
    tarefas: number;
    data: string;
    relatorio: string;
    em: string;
    bytes: number;
  } | null;
  ja_sincronizou_hoje: boolean;
}

export async function getSyncStatus(): Promise<SyncStatus> {
  return json(await apiFetch(`${BASE}/sync`));
}

export async function triggerSync(): Promise<{ ok: boolean; mensagem: string }> {
  return json(await apiFetch(`${BASE}/sync`, { method: "POST" }));
}

// ── Manutenção do roster (editor de equipe) ──
export interface RosterPessoa {
  id: number;
  nome: string;
  cargo: string | null;
  equipe: string | null;
  is_supervisor: boolean;
  ativo: boolean;
  squad: string | null;
  posicao: string | null;
  concluido: number;
  pendente: number;
}

export async function getRoster(team: string): Promise<RosterPessoa[]> {
  const r = await json<{ pessoas: RosterPessoa[] }>(await apiFetch(`${BASE}/roster?team=${team}`));
  return r.pessoas;
}

export async function updateRosterPessoa(
  id: number,
  updates: Partial<{ cargo: string; equipe: string; is_supervisor: boolean; ativo: boolean }>,
): Promise<void> {
  const res = await apiFetch(`${BASE}/roster/${id}`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(updates),
  });
  if (!res.ok) throw new Error(`Erro ${res.status} ao salvar`);
}

export interface Candidato {
  nome: string;
  equipe_atual: string | null;
}

export async function getCandidatos(team: string, busca?: string): Promise<Candidato[]> {
  const qs = new URLSearchParams({ team });
  if (busca) qs.set("busca", busca);
  const r = await json<{ candidatos: Candidato[] }>(await apiFetch(`${BASE}/roster/candidatos?${qs.toString()}`));
  return r.candidatos;
}

export async function adicionarPessoa(nome: string, team: string): Promise<void> {
  const res = await apiFetch(`${BASE}/roster/adicionar`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ nome, team }),
  });
  if (!res.ok) throw new Error(`Erro ${res.status} ao adicionar`);
}

export async function excluirPessoa(id: number): Promise<void> {
  const res = await apiFetch(`${BASE}/roster/${id}`, { method: "DELETE" });
  if (!res.ok) throw new Error(`Erro ${res.status} ao excluir`);
}

File: app/api/v1/endpoints/performance.py
"""Endpoints do módulo "Minha Equipe" (Performance de Equipes).

Restrito a administradores — é monitoramento de desempenho de colaboradores.
Lê das tabelas perf* (populadas pelo seed/ingestão). Ver
`docs/performance-equipes-plano.md`.
"""

import logging
from typing import Optional

from fastapi import APIRouter, BackgroundTasks, Depends, HTTPException, Query, Response, status
from pydantic import BaseModel
from sqlalchemy.orm import Session

from app.core.auth import get_current_user
from app.core.dependencies import get_db
from app.models.legal_one import LegalOneUser
from app.services.performance import relatorios as rel_jobs
from app.services.performance.report import build_individual_pdf, build_sector_pdf
from app.services.performance.service import PerformanceService
from app.services.performance.teams import TEAM_KEYS

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/performance", tags=["Performance (Minha Equipe)"])


def _user_teams(u) -> list:
    return [x for x in (getattr(u, "minha_equipe_equipes", None) or "").split(",") if x]


def _require_minha_equipe(current_user: LegalOneUser = Depends(get_current_user)) -> LegalOneUser:
    """Acesso ao módulo (qualquer time): admin OU permissão do menu."""
    if getattr(current_user, "role", "user") == "admin":
        return current_user
    if not getattr(current_user, "can_use_minha_equipe", False):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Sem permissão para o Minha Equipe.")
    return current_user


def require_team_access(
    team: str = Query(..., description="Slug do time (bb-reu, master-reu, ...)"),
    current_user: LegalOneUser = Depends(get_current_user),
) -> str:
    """Gate por time: admin vê todos; demais precisam do time liberado na árvore."""
    if team not in TEAM_KEYS:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Time inexistente.")
    if getattr(current_user, "role", "user") == "admin":
        return team
    if not getattr(current_user, "can_use_minha_equipe", False):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Sem permissão para o Minha Equipe.")
    if team not in _user_teams(current_user):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Sem acesso a este time.")
    return team


# Alias do gate de módulo (relatórios-job e sync não são por time).
_require_admin = _require_minha_equipe
_admin = Depends(_require_admin)
_team = Depends(require_team_access)


def _require_admin_only(current_user: LegalOneUser = Depends(get_current_user)) -> LegalOneUser:
    """Manutenção do roster é só pra admin (muda atribuição de todos os times)."""
    if getattr(current_user, "role", "user") != "admin":
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Restrito a administradores.")
    return current_user


_adminonly = Depends(_require_admin_only)


@router.get("/equipe", summary="Lista o time com métricas + KPIs", dependencies=[_team])
def equipe(
    team: str = Query(...),
    days: int = Query(30, ge=1, le=365),
    cargo: Optional[str] = Query(None, description="Filtra por cargo"),
    db: Session = Depends(get_db),
):
    return PerformanceService(db).equipe(days=days, cargo=cargo or None, team=team)


@router.get("/cargos", summary="Cargos distintos do time (filtro)", dependencies=[_team])
def cargos(team: str = Query(...), db: Session = Depends(get_db)):
    return {"cargos": PerformanceService(db).cargos(team=team)}


@router.get("/pessoa/{pessoa_id}", summary="Detalhe de uma pessoa (mix + ritmo/ócio)", dependencies=[_team])
def pessoa(
    pessoa_id: int,
    team: str = Query(...),
    days: int = Query(30, ge=1, le=365),
    db: Session = Depends(get_db),
):
    out = PerformanceService(db).pessoa_detalhe(pessoa_id, days=days)
    if out is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Pessoa não encontrada.")
    return out


@router.get("/tipos", summary="Mapa de impacto por subtipo (do time)", dependencies=[_team])
def tipos(team: str = Query(...), days: int = Query(30, ge=1, le=365), db: Session = Depends(get_db)):
    return {"tipos": PerformanceService(db).tipos(days=days, team=team)}


@router.get("/dashboard", summary="Painel do time: vazão, pool/atrasado, jornada, top tipos", dependencies=[_team])
def dashboard(team: str = Query(...), days: int = Query(30, ge=1, le=365), db: Session = Depends(get_db)):
    return PerformanceService(db).dashboard(days=days, team=team)


@router.get("/subtipo-detalhe", summary="Detalhe (capacity) de um subtipo do board", dependencies=[_team])
def subtipo_detalhe(
    team: str = Query(...),
    subtipo: str = Query(...),
    days: int = Query(30, ge=1, le=365),
    db: Session = Depends(get_db),
):
    return PerformanceService(db).subtipo_setor_detalhe(subtipo, days=days, team=team)


@router.get("/duplicadas", summary="Preview de duplicadas (mesma pasta + subtipo) de um tipo", dependencies=[_team])
def duplicadas(team: str = Query(...), subtipo: str = Query(...), db: Session = Depends(get_db)):
    return PerformanceService(db).duplicadas_subtipo(subtipo, team=team)


@router.get("/board-tarefas", summary="Curadoria do board (subtipos exibidos)", dependencies=[_team])
def board_tarefas(team: str = Query(...), db: Session = Depends(get_db)):
    return PerformanceService(db).board_listar(team)


@router.get("/board-tarefas/catalogo", summary="Subtipos disponíveis pra adicionar ao board", dependencies=[_team])
def board_catalogo(team: str = Query(...), busca: str = Query(""), db: Session = Depends(get_db)):
    return {"subtipos": PerformanceService(db).board_catalogo(team, busca)}


class BoardTarefaReq(BaseModel):
    subtipo: str


@router.post("/board-tarefas", summary="Adiciona um subtipo ao board", dependencies=[_team])
def board_adicionar(team: str = Query(...), req: BoardTarefaReq = ..., db: Session = Depends(get_db)):
    return PerformanceService(db).board_adicionar(team, req.subtipo)


@router.delete("/board-tarefas", summary="Remove um subtipo do board", dependencies=[_team])
def board_remover(team: str = Query(...), subtipo: str = Query(...), db: Session = Depends(get_db)):
    return PerformanceService(db).board_remover(team, subtipo)


class CancelarDupReq(BaseModel):
    subtipo: str


@router.post("/duplicadas/cancelar", summary="Inicia cancelamento de duplicadas (varredura LIVE + cancela)", dependencies=[_team])
def cancelar_duplicadas(team: str = Query(...), req: CancelarDupReq = ...):
    from app.services.performance import cancel_duplicadas

    return {"job_id": cancel_duplicadas.iniciar(team, req.subtipo)}


@router.post("/duplicadas/cancelar/abort", summary="Pede pra parar o lote de cancelamento", dependencies=[_team])
def cancelar_duplicadas_abort(team: str = Query(...), job_id: str = Query(...)):
    from app.services.performance import cancel_duplicadas

    return {"ok": cancel_duplicadas.solicitar_abort(job_id)}


@router.get("/duplicadas/cancelar/status", summary="Progresso do lote de cancelamento", dependencies=[_team])
def cancelar_duplicadas_status(team: str = Query(...), job_id: str = Query(...)):
    from app.services.performance import cancel_duplicadas

    st = cancel_duplicadas.status(job_id)
    if not st:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Job não encontrado.")
    return st


@router.get("/export", summary="Exporta xlsx de um recorte do time", dependencies=[_team])
def export(
    team: str = Query(...),
    escopo: str = Query("atrasado", pattern="^(atrasado|pendente|concluido)$"),
    pessoa_id: Optional[int] = Query(None),
    subtipo: Optional[str] = Query(None),
    days: int = Query(30, ge=1, le=365),
    db: Session = Depends(get_db),
):
    data = PerformanceService(db).export_xlsx(escopo=escopo, pessoa_id=pessoa_id, subtipo=subtipo, days=days, team=team)
    fname = f"minha-equipe-{team}-{escopo}.xlsx"
    return Response(
        content=data,
        media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        headers={"Content-Disposition": f'attachment; filename="{fname}"'},
    )


@router.get("/relatorio-setor", summary="Relatório PDF do time (Sonnet + fallback)", dependencies=[_team])
def relatorio_setor(team: str = Query(...), days: int = Query(30, ge=1, le=365), db: Session = Depends(get_db)):
    pdf = build_sector_pdf(db, days=days, team=team)
    return Response(
        content=pdf,
        media_type="application/pdf",
        headers={"Content-Disposition": f'attachment; filename="relatorio-{team}.pdf"'},
    )


@router.get("/pessoa/{pessoa_id}/relatorio", summary="Relatório PDF individual (raio-x + intervenções)", dependencies=[_team])
def relatorio_pessoa(pessoa_id: int, team: str = Query(...), days: int = Query(30, ge=1, le=365), db: Session = Depends(get_db)):
    pdf = build_individual_pdf(db, pessoa_id, days=days)
    if pdf is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Pessoa não encontrada.")
    return Response(
        content=pdf,
        media_type="application/pdf",
        headers={"Content-Disposition": f'attachment; filename="raio-x-pessoa-{pessoa_id}.pdf"'},
    )


# ── Relatórios como JOB persistente (sobrevivem à navegação) ──────────────
class CriarRelatorioReq(BaseModel):
    tipo: str = "setor"  # 'setor' | 'pessoa'
    team: str = "bb-reu"
    days: int = 30
    pessoa_id: Optional[int] = None


@router.post("/relatorios", summary="Dispara a geração de um relatório (job persistente)")
def criar_relatorio(
    req: CriarRelatorioReq,
    background: BackgroundTasks,
    current_user: LegalOneUser = Depends(_require_minha_equipe),
    db: Session = Depends(get_db),
):
    if req.team not in TEAM_KEYS:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Time inexistente.")
    if getattr(current_user, "role", "user") != "admin" and req.team not in _user_teams(current_user):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Sem acesso a este time.")
    tipo = "pessoa" if req.tipo == "pessoa" else "setor"
    rid, label = rel_jobs.criar(
        db, tipo=tipo, days=req.days, pessoa_id=req.pessoa_id, user_id=current_user.id, team=req.team
    )
    background.add_task(rel_jobs.gerar, rid)
    return {"id": rid, "label": label, "status": "processando"}


@router.get("/relatorios", summary="Lista os relatórios do usuário")
def listar_relatorios(
    current_user: LegalOneUser = Depends(_require_admin),
    db: Session = Depends(get_db),
):
    return {"items": rel_jobs.listar(db, current_user.id)}


@router.get("/relatorios/{relatorio_id}/download", summary="Baixa o PDF de um relatório pronto")
def baixar_relatorio(
    relatorio_id: int,
    current_user: LegalOneUser = Depends(_require_admin),
    db: Session = Depends(get_db),
):
    pdf, _label = rel_jobs.get_pdf(db, relatorio_id, current_user.id)
    if pdf is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Relatório não está pronto ou não encontrado.")
    return Response(
        content=pdf,
        media_type="application/pdf",
        headers={"Content-Disposition": f'attachment; filename="relatorio-minha-equipe-{relatorio_id}.pdf"'},
    )


# ── Ingestão dos dados (download do relatório do L1) ──────────────────────
@router.get("/sync", summary="Último sync da ingestão (download do relatório do L1)", dependencies=[_admin])
def sync_status(db: Session = Depends(get_db)):
    from app.services.performance.report_ingest import get_last_sync, ja_sincronizou_hoje

    return {"last_sync": get_last_sync(), "ja_sincronizou_hoje": ja_sincronizou_hoje()}


def _run_sync_bg() -> None:
    from app.db.session import SessionLocal
    from app.services.performance.report_ingest import baixar_e_ingerir

    db = SessionLocal()
    try:
        baixar_e_ingerir(db, force=True)
    except Exception:  # noqa: BLE001
        logger.exception("Minha Equipe: falha na ingestão manual.")
    finally:
        db.close()


@router.post("/sync", summary="Dispara a ingestão agora (baixa o relatório mais recente do L1)", dependencies=[_admin])
def sync_now(background: BackgroundTasks):
    background.add_task(_run_sync_bg)
    return {"ok": True, "mensagem": "Ingestão disparada — baixando o relatório do L1 e atualizando os dados."}


# ── Manutenção do roster (editor de equipe, admin-only) ───────────────────
@router.get("/roster", summary="Roster editável do time (manutenção)", dependencies=[_adminonly])
def roster(team: str = Query(...), db: Session = Depends(get_db)):
    if team not in TEAM_KEYS:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Time inexistente.")
    return {"pessoas": PerformanceService(db).roster(team)}


class RosterUpdateReq(BaseModel):
    cargo: Optional[str] = None
    equipe: Optional[str] = None
    is_supervisor: Optional[bool] = None
    ativo: Optional[bool] = None


@router.patch("/roster/{pessoa_id}", summary="Atualiza uma pessoa do roster", dependencies=[_adminonly])
def update_roster(pessoa_id: int, req: RosterUpdateReq, db: Session = Depends(get_db)):
    if req.equipe is not None and req.equipe not in TEAM_KEYS:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Time inexistente.")
    out = PerformanceService(db).update_pessoa(
        pessoa_id, cargo=req.cargo, equipe=req.equipe,
        is_supervisor=req.is_supervisor, ativo=req.ativo,
    )
    if out is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Pessoa não encontrada.")
    return out


@router.get("/roster/candidatos", summary="Candidatos a adicionar ao time (usuários do L1)", dependencies=[_adminonly])
def roster_candidatos(team: str = Query(...), busca: Optional[str] = Query(None), db: Session = Depends(get_db)):
    if team not in TEAM_KEYS:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Time inexistente.")
    return {"candidatos": PerformanceService(db).candidatos(team, busca=busca)}


class AddPessoaReq(BaseModel):
    nome: str
    team: str


@router.post("/roster/adicionar", summary="Adiciona uma pessoa ao time", dependencies=[_adminonly])
def roster_adicionar(req: AddPessoaReq, db: Session = Depends(get_db)):
    if req.team not in TEAM_KEYS:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Time inexistente.")
    out = PerformanceService(db).adicionar_pessoa(req.nome, req.team)
    if out is None:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Nome inválido.")
    return out


@router.delete("/roster/{pessoa_id}", summary="Exclui a pessoa do sistema (saiu do escritório)", dependencies=[_adminonly])
def roster_excluir(pessoa_id: int, db: Session = Depends(get_db)):
    out = PerformanceService(db).excluir_pessoa(pessoa_id)
    if out is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Pessoa não encontrada.")
    return out
